File: truevfs-comp/truevfs-comp-mmbs/src/main/scala/net/java/truevfs/comp/mmbs/MultiplexingMBeanServer.scala
/*
 * Copyright (C) 2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.comp.mmbs

import MultiplexingMBeanServer._
import collection.JavaConverters._
import java.util._
import javax.management._

private object MultiplexingMBeanServer {

  def apply(mbs: MBeanServer) = {
    if (classLoader == ClassLoader.getSystemClassLoader) mbs
    else new MultiplexingMBeanServer(mbs)
  }

  private def classLoader = getClass.getClassLoader

  object qualifier {
    val key = "CLASS_LOADER"
    val value = classLoader.getClass.getName + '@' +
      (Integer toHexString (System identityHashCode classLoader))
  }

  private def getKeyPropertyList(name: ObjectName) = name.getKeyPropertyList

  private def copyKeyPropertyList(name: ObjectName) =
    new Hashtable[String, String](name.getKeyPropertyList)

  private val keyPropertyList = {
    val name = new ObjectName(":test=foo")
    try {
      val table = getKeyPropertyList(name)
      table remove "test"
      getKeyPropertyList _
    } catch {
      case _: UnsupportedOperationException =>
        copyKeyPropertyList _
    }
  }

  private def apply(name: ObjectName): ObjectName = {
    if (null eq name) return null
    val domain = name.getDomain
    val table = keyPropertyList(name)
    table put (qualifier.key, qualifier.value)
    ObjectName getInstance (domain, table)
  }

  private def unapply(name: ObjectName): ObjectName = {
    if (null eq name) return null
    val domain = name.getDomain
    val table = keyPropertyList(name)
    table remove qualifier.key
    ObjectName getInstance (domain, table)
  }

  private def apply(instance: ObjectInstance): ObjectInstance = {
    if (null eq instance) return null
    new ObjectInstance(apply(instance.getObjectName), instance.getClassName)
  }

  private def unapply(instance: ObjectInstance): ObjectInstance = {
    if (null eq instance) return null
    new ObjectInstance(unapply(instance.getObjectName), instance.getClassName)
  }
}

private final class MultiplexingMBeanServer(mbs: MBeanServer)
extends DecoratingMBeanServer(mbs) with Immutable {

  override def createMBean(className: String, name: ObjectName) =
    unapply(mbs createMBean (className, apply(name)))

  override def createMBean(className: String, name: ObjectName, loaderName: ObjectName) =
    unapply(mbs createMBean (className, apply(name), loaderName))

  override def createMBean(className: String, name: ObjectName, params: Array[AnyRef], signature: Array[String]) =
    unapply(mbs createMBean (className, apply(name), params, signature))

  override def createMBean(className: String, name: ObjectName, loaderName: ObjectName, params: Array[AnyRef], signature: Array[String]) =
    unapply(mbs createMBean (className, apply(name), loaderName, params, signature))

  override def registerMBean(`object`: Any, name: ObjectName) =
    mbs registerMBean (`object`, apply(name))

  override def unregisterMBean(name: ObjectName) {
    mbs unregisterMBean apply(name)
  }

  override def getObjectInstance(name: ObjectName) =
    unapply(mbs getObjectInstance apply(name))

  override def queryMBeans(name: ObjectName, query: QueryExp) =
    ((mbs queryMBeans (apply(name), query)).asScala map unapply).asJava

  override def queryNames(name: ObjectName, query: QueryExp) =
    ((mbs queryNames (apply(name), query)).asScala map unapply).asJava

  override def isRegistered(name: ObjectName) =
    mbs isRegistered apply(name)

  override def getAttribute(name: ObjectName, attribute: String) =
    mbs getAttribute (apply(name), attribute)

  override def getAttributes(name: ObjectName, attributes: Array[String]) =
    mbs getAttributes (apply(name), attributes)

  override def setAttribute(name: ObjectName, attribute: Attribute) {
    mbs setAttribute (apply(name), attribute)
  }

  override def setAttributes(name: ObjectName, attributes: AttributeList) =
    mbs setAttributes (apply(name), attributes)

  override def invoke(name: ObjectName, operationName: String, params: Array[AnyRef], signature: Array[String]) =
    mbs invoke (apply(name), operationName, params, signature)

  override def addNotificationListener(name: ObjectName, listener: NotificationListener, filter: NotificationFilter, handback: Any) {
    mbs addNotificationListener (apply(name), listener, filter, handback)
  }

  override def addNotificationListener(name: ObjectName, listener: ObjectName, filter: NotificationFilter, handback: Any) {
    mbs addNotificationListener (apply(name), listener, filter, handback)
  }

  override def removeNotificationListener(name: ObjectName, listener: ObjectName) {
    mbs removeNotificationListener (apply(name), listener)
  }

  override def removeNotificationListener(name: ObjectName, listener: ObjectName, filter: NotificationFilter, handback: Any) {
    mbs removeNotificationListener (apply(name), listener, filter, handback)
  }

  override def removeNotificationListener(name: ObjectName, listener: NotificationListener) {
    mbs removeNotificationListener (apply(name), listener)
  }

  override def removeNotificationListener(name: ObjectName, listener: NotificationListener, filter: NotificationFilter, handback: Any) {
    mbs removeNotificationListener (apply(name), listener, filter, handback)
  }

  override def getMBeanInfo(name: ObjectName) = mbs getMBeanInfo apply(name)

  override def isInstanceOf(name: ObjectName, className: String) =
    mbs isInstanceOf (apply(name), className)

  @Deprecated
  override def deserialize(name: ObjectName, data: Array[Byte]) =
    mbs deserialize (apply(name), data)

  override def getClassLoaderFor(mbeanName: ObjectName) =
    mbs getClassLoaderFor apply(mbeanName)
}

File: src/main/java/net/java/truecommons3/jmx/mmbs/MultiplexingMBeanServer.java
/*
 * Copyright (C) 2005-2015 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truecommons3.jmx.mmbs;

import java.io.ObjectInputStream;
import java.util.LinkedHashSet;
import java.util.Objects;
import java.util.Set;
import javax.annotation.concurrent.Immutable;
import javax.management.*;
import net.java.truecommons3.jmx.*;
import net.java.truecommons3.shed.HashMaps;

/**
 * @since  TrueCommons 2.3
 * @author Christian Schlichtherle
 */
@Immutable
public final class MultiplexingMBeanServer extends DelegatingMBeanServer {

    private final MBeanServer mbs;
    private final ObjectNameModifier modifier;

    public MultiplexingMBeanServer(final MBeanServer mbs, final ObjectNameModifier modifier) {
        this.mbs = Objects.requireNonNull(mbs);
        this.modifier = Objects.requireNonNull(modifier);
    }

    @Override
    protected MBeanServer mbs() { return mbs; }

    @Override
    public ObjectInstance createMBean(String className, ObjectName name)
    throws ReflectionException, InstanceAlreadyExistsException, MBeanRegistrationException, MBeanException, NotCompliantMBeanException {
        return modifier.unapply(mbs().createMBean(className, modifier.apply(name)));
    }

    @Override
    public ObjectInstance createMBean(String className, ObjectName name, ObjectName loaderName)
    throws ReflectionException, InstanceAlreadyExistsException, MBeanRegistrationException, MBeanException, NotCompliantMBeanException, InstanceNotFoundException {
        return modifier.unapply(mbs().createMBean(className, modifier.apply(name), loaderName));
    }

    @Override
    public ObjectInstance createMBean(String className, ObjectName name, Object[] params, String[] signature)
    throws ReflectionException, InstanceAlreadyExistsException, MBeanRegistrationException, MBeanException, NotCompliantMBeanException {
        return modifier.unapply(mbs().createMBean(className, modifier.apply(name), params, signature));
    }

    @Override
    public ObjectInstance createMBean(String className, ObjectName name, ObjectName loaderName, Object[] params, String[] signature)
    throws ReflectionException, InstanceAlreadyExistsException, MBeanRegistrationException, MBeanException, NotCompliantMBeanException, InstanceNotFoundException {
        return modifier.unapply(mbs().createMBean(className, modifier.apply(name), loaderName, params, signature));
    }

    @Override
    public ObjectInstance registerMBean(Object object, ObjectName name)
    throws InstanceAlreadyExistsException, MBeanRegistrationException, NotCompliantMBeanException {
        return modifier.unapply(mbs().registerMBean(object, modifier.apply(name)));
    }

    @Override
    public void unregisterMBean(ObjectName name)
    throws MBeanRegistrationException, InstanceNotFoundException {
        mbs().unregisterMBean(modifier.apply(name));
    }

    @Override
    public ObjectInstance getObjectInstance(ObjectName name) throws InstanceNotFoundException {
        return modifier.unapply(mbs().getObjectInstance(modifier.apply(name)));
    }

    @Override
    public Set<ObjectInstance> queryMBeans(ObjectName name, QueryExp query) {
        final Set<ObjectInstance> is = mbs().queryMBeans(modifier.apply(name), query);
        final Set<ObjectInstance> os = new LinkedHashSet<>(HashMaps.initialCapacity(is.size()));
        for (ObjectInstance i : is)
            os.add(modifier.unapply(i));
        return os;
    }

    @Override
    public Set<ObjectName> queryNames(ObjectName name, QueryExp query) {
        final Set<ObjectName> is = mbs().queryNames(modifier.apply(name), query);
        final Set<ObjectName> os = new LinkedHashSet<>(HashMaps.initialCapacity(is.size()));
        for (ObjectName i : is)
            os.add(modifier.unapply(i));
        return os;
    }

    @Override
    public boolean isRegistered(ObjectName name) {
        return mbs().isRegistered(modifier.apply(name));
    }

    @Override
    public Integer getMBeanCount() {
        return mbs().getMBeanCount();
    }

    @Override
    public Object getAttribute(ObjectName name, String attribute) throws MBeanException, AttributeNotFoundException, InstanceNotFoundException, ReflectionException {
        return mbs().getAttribute(modifier.apply(name), attribute);
    }

    @Override
    public AttributeList getAttributes(ObjectName name, String[] attributes) throws InstanceNotFoundException, ReflectionException {
        return mbs().getAttributes(modifier.apply(name), attributes);
    }

    @Override
    public void setAttribute(ObjectName name, Attribute attribute) throws InstanceNotFoundException, AttributeNotFoundException, InvalidAttributeValueException, MBeanException, ReflectionException {
        mbs().setAttribute(modifier.apply(name), attribute);
    }

    @Override
    public AttributeList setAttributes(ObjectName name, AttributeList attributes) throws InstanceNotFoundException, ReflectionException {
        return mbs().setAttributes(modifier.apply(name), attributes);
    }

    @Override
    public Object invoke(ObjectName name, String operationName, Object[] params, String[] signature) throws InstanceNotFoundException, MBeanException, ReflectionException {
        return mbs().invoke(modifier.apply(name), operationName, params, signature);
    }

    @Override
    public void addNotificationListener(ObjectName name, NotificationListener listener, NotificationFilter filter, Object handback) throws InstanceNotFoundException {
        mbs().addNotificationListener(modifier.apply(name), listener, filter, handback);
    }

    @Override
    public void addNotificationListener(ObjectName name, ObjectName listener, NotificationFilter filter, Object handback) throws InstanceNotFoundException {
        mbs().addNotificationListener(modifier.apply(name), listener, filter, handback);
    }

    @Override
    public void removeNotificationListener(ObjectName name, ObjectName listener) throws InstanceNotFoundException, ListenerNotFoundException {
        mbs().removeNotificationListener(modifier.apply(name), listener);
    }

    @Override
    public void removeNotificationListener(ObjectName name, ObjectName listener, NotificationFilter filter, Object handback) throws InstanceNotFoundException, ListenerNotFoundException {
        mbs().removeNotificationListener(modifier.apply(name), listener, filter, handback);
    }

    @Override
    public void removeNotificationListener(ObjectName name, NotificationListener listener) throws InstanceNotFoundException, ListenerNotFoundException {
        mbs().removeNotificationListener(modifier.apply(name), listener);
    }

    @Override
    public void removeNotificationListener(ObjectName name, NotificationListener listener, NotificationFilter filter, Object handback) throws InstanceNotFoundException, ListenerNotFoundException {
        mbs().removeNotificationListener(modifier.apply(name), listener, filter, handback);
    }

    @Override
    public MBeanInfo getMBeanInfo(ObjectName name) throws InstanceNotFoundException, IntrospectionException, ReflectionException {
        return mbs().getMBeanInfo(modifier.apply(name));
    }

    @Override
    public boolean isInstanceOf(ObjectName name, String className) throws InstanceNotFoundException {
        return mbs().isInstanceOf(modifier.apply(name), className);
    }

    @Override
    @Deprecated
    public ObjectInputStream deserialize(ObjectName name, byte[] data) throws InstanceNotFoundException, OperationsException {
        return mbs().deserialize(modifier.apply(name), data);
    }

    @Override
    public ClassLoader getClassLoaderFor(ObjectName mbeanName) throws InstanceNotFoundException {
        return mbs().getClassLoaderFor(modifier.apply(mbeanName));
    }
}
